examples/s3_demo.cc

```diff
@@ -1,172 +1,192 @@
-
-          if (i + 1 < num_iters)
-            ThisFiber::SleepFor(chrono::seconds(GetFlag(FLAGS_delay)));
-        }
-        return ec;
-      });
-
-      CHECK(!ec) << ec;
-    } else if (cmd == "read") {
-      pp->GetNextProactor()->Await([&] {
-        auto ec = bucket.Connect(300);
-        CHECK(!ec);
-
-        io::Result<io::ReadonlyFile*> res = bucket.OpenReadFile(obj_path);
-        if (res) {
-          io::ReadonlyFile* file = *res;
-          std::unique_ptr<uint8_t[]> buf(new uint8_t[1024]);
-          io::SizeOrError sz_res = file->Read(0, io::MutableBytes(buf.get(), 1024));
-          if (sz_res) {
-            CONSOLE_INFO << "File contents(first 1024) of " << obj_path << ":";
-            CONSOLE_INFO << string_view(reinterpret_cast<char*>(buf.get()), *sz_res);
-          } else {
-            LOG(ERROR) << "Error: " << sz_res.error();
-          }
-        } else {
-          LOG(ERROR) << "Read Error: " << res.error().message();
-        }
-      });
-    } else if (cmd == "write") {
-      pp->GetNextProactor()->Await([&] {
-        auto ec = bucket.Connect(300);
-        CHECK(!ec);
-
-        io::Result<io::WriteFile*> res = bucket.OpenWriteFile(obj_path);
-        if (res) {
-          unique_ptr<io::WriteFile> file{*res};
-          CHECK(file);
-          std::unique_ptr<uint8_t[]> buf(new uint8_t[1024]);
-          memset(buf.get(), 'R', 1024);
-          for (size_t i = 0; i < GetFlag(FLAGS_write_factor); ++i) {
-            ec = file->Write(io::Bytes(buf.get(), 1024));
-            CHECK(!ec);
-          }
-          ec = file->Close();
-          CHECK(!ec);
-        } else {
-          LOG(ERROR) << "Error: " << res.error();
-        }
-      });
-    } else {
-      LOG(ERROR) << "Unknown command " << cmd;
+    if (*n == 0) {
+      LOG(INFO) << "finished download; read_n=" << read_n;
+      return;
     }
+    read_n += *n;
+  }
+}
+
+int main(int argc, char* argv[]) {
+  MainInitGuard guard(&argc, &argv);
+
+  std::unique_ptr<util::ProactorPool> pp;
+
+#ifdef __linux__
+  if (absl::GetFlag(FLAGS_epoll)) {
+    pp.reset(util::fb2::Pool::Epoll());
+  } else {
+    pp.reset(util::fb2::Pool::IOUring(256));
   }
+#else
+  pp.reset(util::fb2::Pool::Epoll());
+#endif
+
+  pp->Run();
 
-  pp->Stop();
+  pp->GetNextProactor()->Await([&] {
+    util::aws::Init();
+    std::string cmd = absl::GetFlag(FLAGS_cmd);
+    LOG(INFO) << "s3v2_demo; cmd=" << cmd;
+
+    if (cmd == "list-buckets") {
+      ListBuckets();
+    } else if (cmd == "list-objects") {
+      ListObjects(absl::GetFlag(FLAGS_bucket));
+    } else if (cmd == "upload") {
+      Upload(absl::GetFlag(FLAGS_bucket), absl::GetFlag(FLAGS_key),
+             absl::GetFlag(FLAGS_upload_size), absl::GetFlag(FLAGS_chunk_size));
+    } else if (cmd == "download") {
+      Download(absl::GetFlag(FLAGS_bucket), absl::GetFlag(FLAGS_key));
+    } else {
+      LOG(ERROR) << "unknown command: " << cmd;
+    }
 
-  return 0;
+    util::aws::Shutdown();
```

Comment:
you should call `pp->Stop()` and then `return 0` at the end

Reply:
yep sure